lib/puppet/util/tagging.rb

```diff
@@ -40,7 +40,7 @@ def tag(*ary)
   # since that results in testing the same string twice
   #
   def tag_if_valid(name)
-    if name.is_a?(String) && name =~ ValidTagRegex
+    if name.is_a?(String) && name =~ ValidTagRegex && Puppet::Resource::MAIN != name
```

Comment:
This is the special case @joshcooper pointed out - when name is a String and 'main', it is special case to consider - we don't know all the places that may matter.